tests/components/forked_daapd/test_media_player.py

```diff
@@ -0,0 +1,726 @@
+    if additional_service_data is None:
+        additional_service_data = {}
+    return await hass.services.async_call(
+        MP_DOMAIN,
+        service,
+        service_data={ATTR_ENTITY_ID: entity_name, **additional_service_data},
+        blocking=blocking,
+    )
+
+
+async def test_zone(hass, mock_api_object):
+    """Test zone attributes and methods."""
+    zone_entity_name = TEST_ZONE_ENTITY_NAMES[0]
+    state = hass.states.get(zone_entity_name)
+    assert state.attributes[ATTR_FRIENDLY_NAME] == "forked-daapd output (kitchen)"
+    assert state.attributes[ATTR_SUPPORTED_FEATURES] == SUPPORTED_FEATURES_ZONE
+    assert state.state == STATE_ON
+    assert state.attributes[ATTR_MEDIA_VOLUME_LEVEL] == 0.5
+    assert not state.attributes[ATTR_MEDIA_VOLUME_MUTED]
+    await _service_call(hass, zone_entity_name, SERVICE_TURN_ON)
+    await _service_call(hass, zone_entity_name, SERVICE_TURN_OFF)
+    await _service_call(hass, zone_entity_name, SERVICE_TOGGLE)
+    await _service_call(
+        hass, zone_entity_name, SERVICE_VOLUME_SET, {ATTR_MEDIA_VOLUME_LEVEL: 0.3}
+    )
+    await _service_call(
+        hass, zone_entity_name, SERVICE_VOLUME_MUTE, {ATTR_MEDIA_VOLUME_MUTED: True}
+    )
+    await _service_call(
+        hass, zone_entity_name, SERVICE_VOLUME_MUTE, {ATTR_MEDIA_VOLUME_MUTED: False}
+    )
+    zone_entity_name = TEST_ZONE_ENTITY_NAMES[2]
+    await _service_call(hass, zone_entity_name, SERVICE_TOGGLE)
+    await _service_call(
+        hass, zone_entity_name, SERVICE_VOLUME_MUTE, {ATTR_MEDIA_VOLUME_MUTED: True}
+    )
+    output_id = SAMPLE_OUTPUTS_ON[0]["id"]
+    initial_volume = SAMPLE_OUTPUTS_ON[0]["volume"]
+    mock_api_object.change_output.assert_any_call(output_id, selected=True)
+    mock_api_object.change_output.assert_any_call(output_id, selected=False)
+    mock_api_object.set_volume.assert_any_call(output_id=output_id, volume=30)
+    mock_api_object.set_volume.assert_any_call(output_id=output_id, volume=0)
+    mock_api_object.set_volume.assert_any_call(
+        output_id=output_id, volume=initial_volume
+    )
+    output_id = SAMPLE_OUTPUTS_ON[2]["id"]
+    mock_api_object.change_output.assert_any_call(output_id, selected=True)
+
+
+async def test_last_outputs_master(hass, mock_api_object):
+    """Test restoration of _last_outputs."""
+    # Test turning on sends API call
+    await _service_call(hass, TEST_MASTER_ENTITY_NAME, SERVICE_TURN_ON)
+    assert mock_api_object.change_output.call_count == 0
+    assert mock_api_object.set_enabled_outputs.call_count == 1
+    await _service_call(
+        hass, TEST_MASTER_ENTITY_NAME, SERVICE_TURN_OFF
+    )  # should have stored last outputs
+    assert mock_api_object.change_output.call_count == 0
+    assert mock_api_object.set_enabled_outputs.call_count == 2
+    await _service_call(hass, TEST_MASTER_ENTITY_NAME, SERVICE_TURN_ON)
+    assert mock_api_object.change_output.call_count == 3
+    assert mock_api_object.set_enabled_outputs.call_count == 2
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)